src/Strings.h

```diff
@@ -0,0 +1,388 @@
+
+    std::string trimedString(TrimAndReplace(input.begin(), input.end()));
+
+    //wordIterPairVector is used to store the begin and end iterator of words in input
+    std::vector<std::pair<IteratorT, IteratorT>> wordIterPairVector;
+    Details::Parse<IteratorT, IsDelimiterT>(
+        trimedString.begin(),
+        trimedString.end(),
+        isDelimiter,
+        [&wordIterPairVector] (IteratorT &iterBegin, IteratorT &iterEnd) {
+            wordIterPairVector.emplace_back(std::make_pair(iterBegin, iterEnd));
+        }
+    );
+
+    if (wordIterPairVector.size() == 0)
+        throw std::invalid_argument("wordIterPairVector.size() == 0");
+
+    if (ngramRangeMin < 1 || ngramRangeMin > ngramRangeMax )
+        throw std::invalid_argument("ngramRangeMin and ngramRangeMax not valid");
+
+    for (size_t pairIdx = 0; pairIdx < wordIterPairVector.size(); ++pairIdx) {
+        //copy
+        std::string const paddedWord = " " + std::string(wordIterPairVector[pairIdx].first, wordIterPairVector[pairIdx].second) + " ";
+
+        ParseNgramCharHelper<IteratorT>(paddedWord.begin(), paddedWord.end(), ngramRangeMin, ngramRangeMax, callback);
+    }
+}
+
+}
+
+template <typename IteratorT>
+inline std::string ToLower(IteratorT begin,
+                           IteratorT end) {
+    std::string result(begin, end);
+
+#if (defined _MSC_VER)
+#   pragma warning(push)
+#   pragma warning(disable: 4244) // conversion from 'unsigned int' to 'unsigned char', possible loss of data
+#endif
+
+    std::transform(result.begin(), result.end(), result.begin(), [](char c) { return std::tolower(c); });
+
+#if (defined _MSC_VER)
+#   pragma warning(pop)
+#endif
+
+    return result;
+}
+
+template <typename IteratorT>
+inline std::string ToUpper(IteratorT begin,
+                           IteratorT end) {
+    std::string result(begin, end);
+
+#if (defined _MSC_VER)
+#   pragma warning(push)
+#   pragma warning(disable: 4244) // conversion from 'unsigned int' to 'unsigned char', possible loss of data
+#endif
+
+    std::transform(result.begin(), result.end(), result.begin(), [](char c) { return std::toupper(c); });
+
+#if (defined _MSC_VER)
+#   pragma warning(pop)
+#endif
+
+    return result;
+}
+
+template <
+    typename IteratorT,
+    typename PredicateT
+>
+inline std::string TrimLeft(IteratorT begin,
+                            IteratorT end,
+                            PredicateT isPredicate) {
+    while (begin != end) {
+        if (*begin == isPredicate)
+            ++begin;
+        else
+            break;
+    }
+    return std::string(begin, end);
+}
+
+template <
+    typename IteratorT,
+    typename PredicateT
+>
+inline std::string TrimRight(IteratorT begin,
+                             IteratorT end,
+                             PredicateT isPredicate) {
+    IteratorT offset = begin;
+    IteratorT iter = begin;
+    while (iter != end) {
+        if (*iter != isPredicate)
```

Comment:
>*iter != isPredicate [](start = 12, length = 20)

Predicate is usually a function that evaluates the current value. Please, rewire according to the conventions.

Reply:
changed it to a function